Multiple fixes (Nuki ID + README + ACL + DoorSensor/Keypad)

iranl opened the pull request:
- [x] Fix for incorrect Nuki ID creating lots of Home Assistant auto discovery devices/entities
- [x] README update to better reflect ESP32 device support
- [x] Fix for granular ACL support to set admin and full ACL level on first boot
- [x] Remove duplicate `else if` in `NukiWrapper::onConfigUpdateReceived`
- [ ] Show granular ACL correctly on info page in WebCfg
- [ ] Allow forcing DoorSensor/Keypad availability to prevent entity removal from Home Assistant discovery

technyon commented:
- What is the issue with the nuki id? Are you sure this works, because I think after pairing the id must not be changed. Until now the id was fixed, and only changed when a device is unpaired.
- What's the reason behind having to force keypad and the sensor? It sounds a lot like a workaround for a problem that should be fixed in the first place.

iranl commented:
The Nuki ID is a static uint32 that is unique to the Nuki device and is immutable.
It can be found in hexdecimal form in the Nuki app.
It is not to be confused with the Nuki Hub device ID that is randomly generated during pairing.

The Nuki ID value is available through `_nukiConfig.nukiId`.
It is only used by Nuki Hub in the `Nuki(Opener)Wrapper::setupHASS` function to create Home Assistant auto discovery entities.

For some reason (cause unknown at this time) a seemingly valid config is received with an incorrect `nukiId`. 
It seems this behaviour mainly occures around firmware updates of Nuki Hub (you can imagine I do alot of those while testing my PRs). 

This creates alot of duplicate MQTT entries under the `homeassistant` topic with incorrect Nuki IDs. Home Assistant functionality is not really impacted, but it will create lots of duplicate and dangling entities in MQTT and Home Assistant.

By saving the (immutable and unique to a specific Nuki device) ID to preferences once and only running `Nuki(Opener)Wrapper::setupHASS` if the received config matches this ID this behaviour can be mitigated, altough it is no root cause fix.

In regard to forcing the keypad and door sensor: I have noticed that `Nuki(Opener)Wrapper::setupHASS` can be run before Nuki Hub has determined if a keypad or door sensor is available. 

Some kind of race condition seems to be involved here. 
You can see this behaviour in the WebCfg after reboot where it takes a while for settings related to the keypad to show.

If `Nuki(Opener)Wrapper::setupHASS` runs  while Nuki Hub has not determined that a keypad or door sensor is available it will remove entities related to this from Home Assistant. This should also recover at some time, but will lead to a varying amount of time where these entities are not available in Home Assistant.

technyon commented:
Hi,

ok I confused the IDs, should be fine then. Let me have a look at the race condition first. I'd rather have a proper fix than a workaround.

technyon commented:
I have a theory what's happening: In NukiWrapper::updateConfig are these lines:

readConfig();
readAdvancedConfig();
_configRead = true;

_configRead is checked later to determine wether or not to do the HA setup. It is possible that reading the config from the lock fails, but _configRead is set to true unconditionally. Moving it down into the 

if(_nukiConfigValid) { ... }

block could fix the issue. In that case  _configRead will only be set to true if the call to read the config came back successfully.

iranl commented:
`SetupHASS()` function already also has a check for `_nukiConfigValid` though.
But maybe the bogus Nuki ID configs are also interferring here and the fix for that part maybe already fixes this issue.

I've removed the force keypad/doorsensor options for now and moved keypad to the config valid (and Nuki ID valid) block.

technyon commented:
I've thought about this, this dynamically adding and removing maybe is more complicated than it should be. How about we don't automatically remove auto discovery topics, but instead add a function "Clear autodiscovery" that deletes all entries and then restarts the ESP so they can be rebuilt?

iranl commented:
Current way to enable auto discovery already leads to questions from users who expect auto discovery to be enabled "out-of-the-box". Having a function as suggested would also mean it would be mandatory to "clear" when adding a keypad or doorsensor before it will show up in auto discovery. I prefer the current implementation with some added checks.

technyon commented:
No in case of adding devices it can be automatic, I just wouldn't delete auto discovery topic automatically, so they can be reinitialized on the next start.

iranl commented:
With the extra checks introduced in this PR there shouldn't be anymore unneeded deletions of auto discovery topics.

I don't really see the benefit or need in changing this behaviour at this time.
If this PR fails to fix all issues in this regard the suggested solution could be an option in the future imho.

technyon commented:
Alright let's merge it for now then.
